Document/0x06c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1150,6 +1151,136 @@ Reading 2.390625MB ...
 
 To learn more, please refer to the [r2frida wiki](https://github.com/enovella/r2frida-wiki/blob/master/README.md "r2frida Wiki").
 
+##### Information Gathering
+
+In previous section we learned about using radare2 to collect runtime information of a running process. Radare2 is internally using Frida (r2frida) to obtain process information. In this section we will learn how to use Frida directly to gather process related information.
+
+##### Getting Loaded Classes and their Methods
+
+In Frida REPL Objective-C runtime related information can be accessed using `ObjC` command. Within `ObjC` command, `enumerateLoadedClasses` function lists the loaded classes for a given application.
+
+```
+$ frida -U -f com.iOweAss
+
+[iPhone::com.iOweAss]-> ObjC.enumerateLoadedClasses()
+{
+    "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation": [
+        "__NSBlockVariable__",
+        "__NSGlobalBlock__",
+        "__NSFinalizingBlock__",
+        "__NSAutoBlock__",
+        "__NSMallocBlock__",
+        "__NSStackBlock__"
+    ],
+    "/private/var/containers/Bundle/Application/F390A491-3524-40EA-B3F8-6C1FA105A23A/iOweAss.app/iOweAss": [
+        "JailbreakDetection",
+        "CriticalLogic",
+        "ViewController",
+        "AppDelegate"
+    ]
+}
+
+```
+
+Using `ObjC.classes.<classname>.$ownMethods` the methods declared in each class can be listed.
+
+```
+[iPhone::com.iOweAss]-> ObjC.classes.JailbreakDetection.$ownMethods
+[
+    "+ isJailbroken"
+]
+
+[iPhone::com.iOweAss]-> ObjC.classes.CriticalLogic.$ownMethods
+[
+    "+ doSha256:",
+    "- a:",
+    "- AES128Operation:data:key:iv:",
+    "- coreLogic",
+    "- bat",
+    "- b:",
+    "- hexString:"
+]
+```
+
+##### Getting Loaded Libraries
+
+In Frida REPL process related information ca be obtained using `Process` command. Within `Process` command, `enumerateModules` function lists the libraries loaded into process memory.
+
+```
+[iPhone::com.iOweAss]-> Process.enumerateModules()
+[
+    {
+        "base": "0x10008c000",
+        "name": "iOweAss",
+        "path": "/private/var/containers/Bundle/Application/F390A491-3524-40EA-B3F8-6C1FA105A23A/iOweAss.app/iOweAss",
+        "size": 49152
+    },
+    {
+        "base": "0x1a1c82000",
+        "name": "Foundation",
+        "path": "/System/Library/Frameworks/Foundation.framework/Foundation",
+        "size": 2859008
+    },
+    {
+        "base": "0x1a16f4000",
+        "name": "libobjc.A.dylib",
+        "path": "/usr/lib/libobjc.A.dylib",
+        "size": 200704
+    },
+
+    ...
+```
+
+Similarly, information related to various threads can be obtained.
+
+```
+Process.enumerateThreads()
+[
+    {
+        "context": {
+            "fp": "0x16fd72b90",
+            "lr": "0x1a17a7aa0",
+            "pc": "0x1a17a8634",
```

Comment:
Can you please remove all lines from here till "x9"? To make the output more readable.
Just replace it with "..."

Reply:
Done